presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftHiveMetastore.java

```diff
@@ -154,6 +162,79 @@ public ThriftHiveMetastoreStats getStats()
+                    .stopOnIllegalExceptions()
+                    .run("getPrimaryKey", stats.getGetPrimaryKey().wrap(() ->
+                            getMetastoreClientThenCall(metastoreContext, client -> client.getPrimaryKey(dbName, tableName))));
+
+            if (!pkResponse.isPresent() || pkResponse.get().getPrimaryKeys().size() == 0) {
+                return Optional.empty();
+            }
+
+            List<SQLPrimaryKey> pkCols = pkResponse.get().getPrimaryKeys();
+            boolean isEnabled = pkCols.get(0).isEnable_cstr();
+            boolean isRely = pkCols.get(0).isRely_cstr();
+            String pkName = pkCols.get(0).getPk_name();
+            Set<String> keyCols = pkCols.stream().map(pk -> pk.getColumn_name()).collect(Collectors.toSet());
+            return Optional.of(new PrimaryKeyConstraint<>(pkName, keyCols, isEnabled, isRely));
+        }
+        catch (TException e) {
+            throw new PrestoException(HIVE_METASTORE_ERROR, e);
+        }
+        catch (Exception e) {
+            throw propagate(e);
+        }
+    }
+
+    @Override
+    public List<UniqueConstraint<String>> getUniqueConstraints(MetastoreContext metastoreContext, String dbName, String tableName)
+    {
+        try {
+            Optional<UniqueConstraintsResponse> uniqueConstraintsResponse = retry()
+                    .stopOnIllegalExceptions()
+                    .run("getUniqueConstraints", stats.getGetUniqueConstraints().wrap(() ->
+                            getMetastoreClientThenCall(metastoreContext, client -> client.getUniqueConstraints("hive", dbName, tableName))));
+
+            if (!uniqueConstraintsResponse.isPresent() || uniqueConstraintsResponse.get().getUniqueConstraints().size() == 0) {
+                return ImmutableList.of();
+            }
+
+            List<SQLUniqueConstraint> uniqueCols = uniqueConstraintsResponse.get().getUniqueConstraints();
+            //bucket the unique constraint columns by constraint name
+            Map<String, List<SQLUniqueConstraint>> bucketedConstraints = new HashMap<>();
+            uniqueCols.stream().forEach(uc -> {
+                String constraintName = uc.getUk_name();
+                if (bucketedConstraints.containsKey(constraintName)) {
+                    bucketedConstraints.get(constraintName).add(uc);
+                }
+                else {
+                    List<SQLUniqueConstraint> uniqueColumns = new ArrayList<>();
+                    uniqueColumns.add(uc);
+                    bucketedConstraints.put(constraintName, uniqueColumns);
+                }
+            });
+            //create a unique table constraint per bucket
+            List<UniqueConstraint<String>> result = new ArrayList<>();
+            bucketedConstraints.entrySet().stream().forEach(e -> {
+                String constraintName = e.getKey();
+                Set<String> columnNames = e.getValue().stream().map(c -> c.getColumn_name()).collect(Collectors.toSet());
```

Comment:
`.map(c::getColumn_name).collect(toImmutableSet())`

Reply:
Addressed in latest changes.